scripts/pre_commit_linter.py

```diff
@@ -1927,6 +1933,99 @@ def _check_for_copyright_notice(self):
 
         return summary_messages
 
+    def _check_codeowner_file(self):
+        """Checks the CODEOWNERS file for any uncovered dirs/files and also
+        checks that every pattern in the CODEOWNERS file matches at least one
+        file/dir. Note that this checks the CODEOWNERS file according to the
+        glob patterns supported by Python2.7 environment. For more information
+        please refer https://docs.python.org/2/library/glob.html.
+        """
+        if self.verbose_mode_enabled:
+            print 'Starting CODEOWNERS file check'
+            print '----------------------------------------'
+
+        with _redirect_stdout(_TARGET_STDOUT):
+            codeowner_file = '.github/CODEOWNERS'
+            failed = False
+            summary_messages = []
+            # Checks whether every pattern in the CODEOWNERS file matches at
+            # least one dir/file.
+            path_patterns = []
+            for line_num, line in enumerate(FileCache.readlines(
+                    codeowner_file)):
+                if line.strip() != '' and line.strip()[0] != '#':
+                    if '@' not in line:
+                        print ('%s --> Pattern on line %s doesn\'t have'
+                               'codeowner' % (codeowner_file, line_num))
+                        failed = True
+                    else:
+                        line_in_concern = line.split('@')[0].strip()
```

Comment:
The logic here and below is too complicated to be understood at a glance. Please put comments before all important steps, explaining what is now being done by the line below.

Reply:
Added some comments. PTAL if they look good!